presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java

```diff
@@ -0,0 +1,133 @@
+import static com.teradata.prestomanager.controller.NodeSet.Agent.NodeRole.WORKER;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class NodeSet
+{
+    private static NodeSet nodeSetInstance = new NodeSet();
+
+    public static Set<Agent> nodeSet = ConcurrentHashMap.newKeySet();
+
+    public static NodeSet getInstance()
+    {
+        return nodeSetInstance;
+    }
+
+    public void addAgent(URI uri, NodeRole nodeRole)
+    {
+        nodeSet.add(new Agent(uri, nodeRole));
+    }
+
+    public void removeAgent(int id)
+    {
+        nodeSet.removeIf(agent -> agent.getId().equals(id));
+    }
+
+    private URI getUriById(int id)
+    {
+        return nodeSet.stream().filter(agent -> agent.getId().equals(id))
+                .map(Agent::getUri).findAny().orElseThrow(() -> new IllegalArgumentException("Invalid node ID"));
+    }
+
+    public Collection<URI> getUrisByIds(Collection<Integer> ids)
+    {
+        Collection<URI> uriList = new ArrayList<>();
+        for (int id : ids) {
+            uriList.add(getUriById(id));
+        }
+        return uriList;
+    }
+
+    public Collection<URI> getUris(ApiScope scope)
+    {
+        switch (scope) {
+            case CLUSTER:
+                return nodeSet.stream().map(Agent::getUri).collect(Collectors.toList());
+            case COORDINATOR:
+                return nodeSet.stream().filter(agent -> agent.getNodeRole() == COORDINATOR)
+                        .map(Agent::getUri).collect(Collectors.toList());
+            case WORKERS:
+                return nodeSet.stream().filter(agent -> agent.getNodeRole() == WORKER)
+                        .map(Agent::getUri).collect(Collectors.toList());
+            default:
+                return null;
+        }
+    }
+
+    //TODO: Add method to generate ID
+    public static class Agent
+    {
+        private URI uri;
+        private static final AtomicInteger count = new AtomicInteger(0);
+        private int id;
+        private NodeRole nodeRole;
+
+        public enum NodeRole
+        {
+            WORKER, COORDINATOR
+        }
+
+        private Agent(URI uri, NodeRole nodeRole)
+        {
+            this.uri = requireNonNull(uri, "uri is null");
+            this.nodeRole = requireNonNull(nodeRole, "nodeRole is null");
+
+            this.id = count.incrementAndGet();
+        }
+
+        private URI getUri()
+        {
+            return uri;
+        }
+
+        private Integer getId()
+        {
+            return id;
+        }
+
+        private NodeRole getNodeRole() {return nodeRole;}
+
+        @Override
+        public boolean equals(Object obj)
+        {
+            return (obj instanceof Agent) && (getId().equals(((Agent) obj).getId()))
+                    && Objects.equals(this.getUri(), ((Agent) obj).getUri())
+                    && Objects.equals(this.getNodeRole(), ((Agent) obj).getNodeRole());
```

Comment:
Agents should compare for equals by their ID; this ensures that an ID will never be duplicated. In this form, multiple nodes may have the same ID, and we never check for that, which may have unexpected consequences.